Require compressed extensions appear in the same relative order. (#455)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

~~~

OuterExtensions consists of one or more ExtensionType values, each of which
reference an extension in ClientHelloOuter. The extensions in OuterExtensions
MUST appear in ClientHelloOuter in the same relative order, however, there is

> no requirement that they be continguous. For examples, OuterExtensions may

Reviewer: ```suggestion
no requirement that they be continguous. For example, OuterExtensions may
```

contain extensions A, B, C, while ClientHelloOuter contains extensions A, D, B,
C, E, F.

The "ech_outer_extensions" extension is only used for compressing the

> ClientHelloInner. It MUST NOT be sent in either ClientHelloOuter or
> ClientHelloInner.

Reviewer: ```suggestion
ClientHelloInner. It can only be included in EncodedClientHelloInner, and MUST
NOT be sent in either ClientHelloOuter or ClientHelloInner.
```

Reviewer: I was really confused by this at first, so hopefully this helps clarify for others.

abort the connection with an "illegal_parameter" alert if any of the following
are true:

* Any referenced extension is missing in ClientHelloOuter.

* "encrypted_client_hello" appears in OuterExtensions.

* OuterExtensions contains duplicate values.

* The extensions in ClientHelloOuter corresponding to those in OuterExtensions
  do not occur in the same order.

If the cost to decode an EncodedClientHelloInner is disproportionately large in
comparison to the input size, a malicious client could exploit this in a denial
of service attack. Implementations are RECOMMENDED to use the following
decoding procedure, which runs in linear time:

1. Let `i` be zero and `n` be the number of extensions in ClientHelloOuter.

1. For each extension type, `ext`, in OuterExtensions:

   * If `ext` is "encrypted_client_hello", abort the connection with an
     "illegal_parameter" alert and terminate this procedure.

   * While the `i` is less than `n` and the `i`th extension of
     ClientHelloOuter does not have type `ext`, increment `i`.

   * If `i` is equal to `n`, abort the connection with an "illegal_parameter"
     alert and terminate this procedure.

>    * Otherwise, the `i`th extension of ClientHelloOuter has type `ext`. Copy

Reviewer 2: One nitpick: "`i`th" renders really badly in text.  Consider just avoiding the backticks.

encrypted SNI induces no additional round trip and operates below the
application layer.

# Linear-time Outer Extension Processing {#linear-outer-extensions}

The following procedure processes the "ech_outer_extensions" extension (see
{{encoding-inner}}) in linear time:

1. Let I be zero and N be the number of extensions in ClientHelloOuter.

1. For each extension type, E, in OuterExtensions:

   * If E is "encrypted_client_hello", abort the connection with an
     "illegal_parameter" alert and terminate this procedure.

>    * While the I is less than N and the Ith extension of

Reviewer: ```suggestion
   * While I is less than N and the I-th extension of
```

     ClientHelloOuter does not have type E, increment I.

   * If I is equal to N, abort the connection with an "illegal_parameter"
     alert and terminate this procedure.

>    * Otherwise, the Ith extension of ClientHelloOuter has type E. Copy

Reviewer: ```suggestion
   * Otherwise, the I-th extension of ClientHelloOuter has type E. Copy
```

     it to the EncodedClientHelloInner and increment I.

> Implementations that retain the ClientHelloOuter extension list in another
> iterable form can equivalently replace I with an iterator.
> 

Reviewer: ```suggestion
```
This seems really language and library specific, so I might just drop it.
